patches/webrtc/fix_fallback_to_x11_capturer_on_wayland.patch

```diff
@@ -0,0 +1,58 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: VerteDinde <keeleymhammond@gmail.com>
+Date: Sun, 5 Mar 2023 21:04:37 -0800
+Subject: fix: fallback to X11 capturer on Wayland
+
+CL: https://webrtc-review.googlesource.com/c/src/+/279163
+
+Desktop Capturer behaves inconsistently on Wayland. PipeWire does not
+always successfully start; if it does not, we return a nullptr rather
+than falling back on the X11 capturer, crashing the application.
+
+If the X11 capturer is enabled, we should at minimum try to fallback
+to X11 for desktop capturer. This patch re-enables that fallback,
+which was previously default behavior.
+
+This patch can be removed when 1) this fix is upstreamed, or 2) the
+stability of PipeWire initialization is improved upstream.
+
+Patch_Filename: fix_fallback_to_x11_desktop_capturer_wayland.patch
+
+diff --git a/modules/desktop_capture/screen_capturer_linux.cc b/modules/desktop_capture/screen_capturer_linux.cc
+index 44993837e8bbd84a11ec9d187349a95f83258910..cd9f8b0be6a19d057fe9150382fb72bc4032b343 100644
+--- a/modules/desktop_capture/screen_capturer_linux.cc
++++ b/modules/desktop_capture/screen_capturer_linux.cc
+@@ -34,11 +34,10 @@ std::unique_ptr<DesktopCapturer> DesktopCapturer::CreateRawScreenCapturer(
+ #endif  // defined(WEBRTC_USE_PIPEWIRE)
+ 
+ #if defined(WEBRTC_USE_X11)
+-  if (!DesktopCapturer::IsRunningUnderWayland())
+-    return ScreenCapturerX11::CreateRawScreenCapturer(options);
+-#endif  // defined(WEBRTC_USE_X11)
+-
++  return ScreenCapturerX11::CreateRawScreenCapturer(options);
```

Comment:
**Warning:** drive by comment.

Does this mean that if the Pipewire code-path fails, this falls back to capturing XWayland windows only (when running in a Wayland session)? Is that a sensible thing to do?

Personally, from an application developer's point of view, I'd rather just have the whole thing fail (at which point the application can ask the user to fix the Pipewire setup) rather than **silently** fallback to **only capturing XWayland windows**.

Or, at the very least, the capture API should provide some feedback to the application that it's using a fallback mode so the application can decide what to do next (i.e.: either continue or give up).

Reply:
@vially I agree it's not ideal, but not falling back to the X11 capturer is a change from previous behavior - when I tested on Ubuntu 20.04 and 22.04, WebRTC curiously often used the X11 capturer rather than the expected PipeWire capturer in previous versions of Electron, even when running on Wayland. With more users migrating to Ubuntu 22.04 (which I believe uses Wayland by default now), we're seeing a high number of reports of desktop capturer nor working. IMO, I think we should try to restore this default behavior while we debug exactly why PipeWire is failing on Wayland.

I like your suggestion about the capture API providing feedback to the application! Will need to think through what that could look like, but it's a good idea - I'd like to prioritize getting this fix in, since it's fixing a hard crash, but I'll implement that logic in a follow up PR.

Reply:
@deepak1556 Oh, that's a good callout, thanks! I originally didn't pass those through. I retested yesterday/this morning on Ubuntu 20.04 & 22.04 and Fedora 35, using those flags (here's a copy of the gist I was using, just for verification: https://gist.github.com/6dc823943ec22be638633768df29169e). Curiously, the X11 capturer was still what we were falling back to - even on Fedora, which I would solidly expect not to 🤔 

I noticed @vially has an open PR here that may help with this?: https://github.com/electron/electron/pull/35630 I'm going to open a tracking issue for this, so I can figure out why we're not using the correct capturer on Wayland.

Reply:
> I'd like to prioritize getting this fix in, since it's fixing a hard crash, but I'll implement that logic in a follow up PR

This sounds very reasonable to me. A lot of the things involved in this issue (Pipewire/Wayland/xdg-desktop-portal/Chromium/Electron) have all seen quite a lot of changes lately and it's hard to keep track of what is and isn't supposed to work in every version/release of each of them.

It may be worth starting the testing from the one most likely to work and go from there. For example:
- is *any* native Wayland app able to capture the desktop on Ubuntu 22.04? GTK or QT based desktop capturers should be good candidates for testing this.
- is *Chromium* (when using the same version that's being used in Electron) able to capture the desktop on Ubuntu 22.04? The "WebRTC PipeWire support" and "Preferred Ozone platform" need to be manually configured in `chrome://flags` when testing this or the capturer will fallback to the X11 code path.

The answer to these questions should give a pretty good idea whether this is indeed an upstream bug (PipeWire, Chromium, xdg-desktop-portal, etc) or an Electron specific issue.

I really want Electron to have *great* Wayland support and that's why I sometimes get involved in these discussions. But as I've mentioned before, I'm just leaving drive-by comments here so feel free to completely ignore them.